File: Programmers/Level2/가장 큰 수.py
def solution(numbers):
    answer = merge_sort(numbers)

    return ''.join(answer) if answer[0] != '0' else '0'


def merge_sort(array):
    if len(array) == 1:
        return array
    mid = len(array) // 2
    left = array[:mid]
    right = array[mid:]
    return merge(merge_sort(left), merge_sort(right))


def merge(left, right):
    result = []
    left = [str(i) for i in left]
    right = [str(i) for i in right]
    while left and right:
        if (left[0] + right[0]) > (right[0] + left[0]):
            result.append(left.pop(0))
        else:
            result.append(right.pop(0))
    while left:
        result.append(left.pop(0))
    while right:
        result.append(right.pop(0))
    return result

print(solution([3, 30, 34, 5, 9]))

File: 2021/Programmers/정렬/가장_큰_수.js
const merge = (left, right) => {
  const result = [];
  left = left.map((v) => String(v));
  right = right.map((v) => String(v));
  while (left.length > 0 && right.length > 0) {
    if (left[0] + right[0] > right[0] + left[0]) result.push(left.shift());
    else result.push(right.shift());
  }
  return [...result, ...left, ...right];
};

const merge_sort = (list) => {
  if (list.length === 1) return list;
  const mid = Math.floor(list.length / 2);
  const left = list.slice(0, mid);
  const right = list.slice(mid);
  return merge(merge_sort(left), merge_sort(right));
};

const solution = (numbers) => {
  let list = merge_sort(numbers);
  return list[0] === "0" ? "0" : list.join("");
};

console.log(solution([6, 10, 2]));
